### PR TITLE
Separate signal monitors from runners/plugins

### PR DESCRIPTION
**Pull request by mvd-ows**
This change is [Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)

### CODE REVIEW

File: src/monitor/other.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+use super::*;
+
+pub fn create_signal_monitor() -> impl Future<Item = (), Error = io::Error> + Send {
+    // FIXME: What can I do ...
+    // Blocks forever
+    futures::empty::<(), ()>().and_then(|_| {})
```
Comment (faern) [inline src/monitor/other.rs]: I'm not sure the `.and_then(|_| {})` part is needed? Looks like you can make a future that fits the expected signature directly from `futures::empty::<(), io::Error>()`?

**Review by faern — APPROVED** [review]
Looks really good! Nice step towards being able to use as a library.

File: src/monitor/other.rs (hunk 1 above)
Comment (mvd-ows) [reply src/monitor/other.rs]: Would that give the same behavior? The future isn't supposed to ever complete.

Comment (faern) [reply src/monitor/other.rs]: Yes. The docs for `futures::empty` says "A future which is never resolved.". So `futures::empty::<T, E>()` would return a future that never completes and that implements `Future<Item=T, Error=E>`.

**Review by mvd-ows — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: 0 of 10 files reviewed, 1 unresolved discussion (waiting on @faern)

___
*[src/monitor/other.rs, line 6 at r1](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LTXy4HE-rw719tkMFvu-r1-6:-LTXy7jg7DOtPp6IGu7a:b-nnfkl) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/b6e20c1ff3f927459420c3e61d134bfda081ee01/src/monitor/other.rs#L6)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Yes. The docs for `futures::empty` says "A future which is never resolved.". So `futures::empty::<T, E>()` would return a future that never completes and that implements `Future<Item=T, Error=E>`.
</blockquote></details>

Excellent. I'll update the code with your suggestion.

**Review by faern — CHANGES_REQUESTED** [review]
Since my PRs got merge `master` moved. So you'll have to rebase this PR.

> Reviewed 6 of 10 files at r1, 2 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-:-LTc52m84Ci_IBG28AzX:by3jvji)* status: 8 of 12 files reviewed, 1 unresolved discussion (waiting on @mvd-ows and @faern)

___
*[src/plugin/monitor.rs, line 6 at r2](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LTc6yjg00iXaHa21_4I:-LTc6yjg00iXaHa21_4J:b-wukpfz) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/d4025f79a1181db7822c63420ca4338bcc6d2307/src/plugin/monitor.rs#L6)):*
> ```Rust
> use std::{io, thread, time::Duration};
> 
> pub fn create_plugin_monitor(plugins: Vec<Plugin>) -> impl Future<Item = (), Error = io::Error> + Send {
> ```

It looks like currently this creates a `Future` that will wait for all plugins and if one plugin dies it will kill the rest as well. But there is no way to terminate the plugins without one of them dying. So if ctrl-c is sent, the plugins will never be instructed to exit?

**Review by faern — CHANGES_REQUESTED** [review]
> Reviewed 2 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: 8 of 14 files reviewed, 3 unresolved discussions (waiting on @faern and @mvd-ows)

___
*[src/monitor/mod.rs, line 1 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU4t43z-M8Jx3IunEGh:-LU4t43z-M8Jx3IunEGi:bwlnz5w) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/monitor/mod.rs#L1)):*
> ```Rust
> use std::io;
> ```

This import seems unused?
___
*[src/monitor/unix.rs, line 1 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU4ufXR5vDB8ln0by7w:-LU4ufXR5vDB8ln0by7x:bw7h4zm) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/monitor/unix.rs#L1)):*
> ```Rust
> use super::*;
> ```

It seems that the only thing it uses from the parent is `std::io`. Would look nicer to remove `use std::io;` from the parent and just import it directly in the submodules that need it.
___
*[src/monitor/unix.rs, line 9 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU4sH8V9EPSqW4X3phD:-LU4sH8V9EPSqW4X3phE:b-xbiva5) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/monitor/unix.rs#L9)):*
> ```Rust
> 
> pub fn create_signal_monitor() -> impl Future<Item = (), Error = io::Error> + Send {
>     use tokio_signal::unix::Signal;
> ```

Nit, but this use could be moved out to the parent scope. Now the entire file is dedicated to *nix and there is no conditional compilation involved any longer. I think it was in here from the start to avoid unused import warnings, which is not a problem with this design any more.

**Review by faern — COMMENTED** [review]
> Reviewed 1 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: 9 of 14 files reviewed, 5 unresolved discussions (waiting on @faern and @mvd-ows)

___
*[src/monitor/windows.rs, line 5 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU51Gmu-q09ODBrQM5J:-LU51Gmu-q09ODBrQM5K:btxuv73) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/monitor/windows.rs#L5)):*
> ```Rust
> 
> pub fn create_signal_monitor() -> impl Future<Item = (), Error = io::Error> + Send {
>     use tokio_signal::windows::Event;
> ```

Nit, same as *nix. No need to have the import inside the function any longer.
___
*[src/monitor/windows.rs, line 34 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU51XQm0GyEBirpGGos:-LU51XQm0GyEBirpGGot:b71fjoz) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/monitor/windows.rs#L34)):*
> ```Rust
>     });
> 
>     events
> ```

Nit, could slightly simplify this by removing this and `let events =` and also the `;` at the end. Just return the `futures::lazy(...)` expression directly.

**Review by faern — CHANGES_REQUESTED** [review]
> Reviewed 1 of 6 files at r2, 2 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: 11 of 14 files reviewed, 6 unresolved discussions (waiting on @faern and @mvd-ows)

___
*[src/plugin/monitor.rs, line 24 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU54gv953PJ-PJhujsm:-LU54gv953PJ-PJhujsn:brrhqdl) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/plugin/monitor.rs#L24)):*
> ```Rust
>             match abort_signal.load(Ordering::Relaxed) {
>                 true => Ok(Async::Ready(())),
>                 false => Ok(Async::NotReady),
> ```

I'm not sure this works, have you tried it? If a future returns `Async::NotReady` it will not be polled again until an underlying task reports that it has some event to process. So when implementing the `poll` method on a future one usually need to register some async task with the underlying executor before returning `Async::NotReady`, otherwise it might be stuck forever, never being called.

Or even worse sometimes, it might resolve instantly. If it registers no tasks with the executor the executor might think there is no work left to perform and shut down.

I was stuck with this just yesterday, which is why it's fresh in my memory :D

Not sure exactly what would be the cleanest solution, but not even calling `create_plugin_monitor` if there are no plugins would be nice, because then the resulting process would not need to sit polling some atomic for no reason. And I guess not using any plugins is kind of the default case, so makes sense to slightly optimize for that case?

**Review by mvd-ows — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: 11 of 14 files reviewed, 5 unresolved discussions (waiting on @faern and @mvd-ows)

___
*[src/monitor/windows.rs, line 34 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU51XQm0GyEBirpGGos:-LU5N9Wl5KW_T4_1ZHYQ:b-849dyp) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/monitor/windows.rs#L34)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Nit, could slightly simplify this by removing this and `let events =` and also the `;` at the end. Just return the `futures::lazy(...)` expression directly.
</blockquote></details>

The reason I created a variable is it makes it easier to read IMO. If it's plain wrong or "unrusty" I can of course revise the code.
___
*[src/plugin/monitor.rs, line 6 at r2](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LTc6yjg00iXaHa21_4I:-LU5PLUmCHiVjAri41_T:b-evv1mh) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/d4025f79a1181db7822c63420ca4338bcc6d2307/src/plugin/monitor.rs#L6)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

It looks like currently this creates a `Future` that will wait for all plugins and if one plugin dies it will kill the rest as well. But there is no way to terminate the plugins without one of them dying. So if ctrl-c is sent, the plugins will never be instructed to exit?
</blockquote></details>

I believe it should work. Note that there's only a single abort signal that's being cloned. The same signal is being referenced by both caller or `create_plugin_monitor`, the threads and the future towards the end of this code section.

The caller of `create_plugin_monitor` sets the signal to `true` to shut the monitor down.
___
*[src/plugin/monitor.rs, line 24 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU54gv953PJ-PJhujsm:-LU5O3GY-DwmL48umzkg:bb83seu) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/plugin/monitor.rs#L24)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I'm not sure this works, have you tried it? If a future returns `Async::NotReady` it will not be polled again until an underlying task reports that it has some event to process. So when implementing the `poll` method on a future one usually need to register some async task with the underlying executor before returning `Async::NotReady`, otherwise it might be stuck forever, never being called.

Or even worse sometimes, it might resolve instantly. If it registers no tasks with the executor the executor might think there is no work left to perform and shut down.

I was stuck with this just yesterday, which is why it's fresh in my memory :D

Not sure exactly what would be the cleanest solution, but not even calling `create_plugin_monitor` if there are no plugins would be nice, because then the resulting process would not need to sit polling some atomic for no reason. And I guess not using any plugins is kind of the default case, so makes sense to slightly optimize for that case?
</blockquote></details>

It kind of works because it's being polled, but it's far from optimal since it's probably not rescheduled when it needs to be.

You're correct in that not creating a monitor would be better in this case.

**Review by faern — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: 11 of 14 files reviewed, 4 unresolved discussions (waiting on @faern and @mvd-ows)

___
*[src/monitor/windows.rs, line 34 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU51XQm0GyEBirpGGos:-LU5RzKMCR3t8MiswrDJ:b-ucdzsg) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/monitor/windows.rs#L34)):*
<details><summary><i>Previously, mvd-ows wrote…</i></summary><blockquote>

The reason I created a variable is it makes it easier to read IMO. If it's plain wrong or "unrusty" I can of course revise the code.
</blockquote></details>

Nothing wrong! If you prefer this style, keep it. It's technically just as good as what I suggested :)

**Review by mvd-ows — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: 6 of 14 files reviewed, 2 unresolved discussions (waiting on @faern)

___
*[src/monitor/mod.rs, line 1 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU4t43z-M8Jx3IunEGh:-LU7NkqT1CwP_hGgdOmF:b-896fix) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/monitor/mod.rs#L1)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

This import seems unused?
</blockquote></details>

Done.
___
*[src/monitor/unix.rs, line 1 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU4ufXR5vDB8ln0by7w:-LU7NpUxETZ3831uC2dv:b-896fix) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/monitor/unix.rs#L1)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

It seems that the only thing it uses from the parent is `std::io`. Would look nicer to remove `use std::io;` from the parent and just import it directly in the submodules that need it.
</blockquote></details>

Done.
___
*[src/monitor/unix.rs, line 9 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU4sH8V9EPSqW4X3phD:-LU7NrBgAg_l85N-3Oeq:b-896fix) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/monitor/unix.rs#L9)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Nit, but this use could be moved out to the parent scope. Now the entire file is dedicated to *nix and there is no conditional compilation involved any longer. I think it was in here from the start to avoid unused import warnings, which is not a problem with this design any more.
</blockquote></details>

Done.
___
*[src/monitor/windows.rs, line 5 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU51Gmu-q09ODBrQM5J:-LU7NzTEBfOnshCXUvuK:b-896fix) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/monitor/windows.rs#L5)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Nit, same as *nix. No need to have the import inside the function any longer.
</blockquote></details>

Done.
___
*[src/plugin/monitor.rs, line 24 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU54gv953PJ-PJhujsm:-LU7O4-dCHKU7Qx9W_gN:b-prhdz0) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/plugin/monitor.rs#L24)):*
<details><summary><i>Previously, mvd-ows wrote…</i></summary><blockquote>

It kind of works because it's being polled, but it's far from optimal since it's probably not rescheduled when it needs to be.

You're correct in that not creating a monitor would be better in this case.
</blockquote></details>

Changed it to panic if there are zero plugins, and adjusted the logic at the call site.

**Review by faern — COMMENTED** [review]
> Reviewed 1 of 8 files at r3, 4 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: 11 of 14 files reviewed, 1 unresolved discussion (waiting on @faern and @mvd-ows)

___
*[src/plugin/monitor.rs, line 24 at r3](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU54gv953PJ-PJhujsm:-LU9qhOz8d6rBGiuLchS:buknrl3) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/3bc82144450df4713a931f1d692eb25f0aef16e7/src/plugin/monitor.rs#L24)):*
<details><summary><i>Previously, mvd-ows wrote…</i></summary><blockquote>

Changed it to panic if there are zero plugins, and adjusted the logic at the call site.
</blockquote></details>

That sounds like a good idea!
___
*[src/relay/local.rs, line 68 at r4](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU9qwgD-dc_ignUN9zG:-LU9qwgD-dc_ignUN9zH:bvhpv8q) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/b0d4612fd9808a9737b92bc1b4c49b0741b4375a/src/relay/local.rs#L68)):*
> ```Rust
> 
>     match plugins.is_empty() {
>         true => {
> ```

All the code in the branches could be one level less indented if we do `if plugin.is_empty() {} else {}` instead of a match. This works as well of course, but I think it's more common to just use `if` if the expression is simply a boolean.

**Review by mvd-ows — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: 11 of 14 files reviewed, 1 unresolved discussion (waiting on @faern)

___
*[src/relay/local.rs, line 68 at r4](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU9qwgD-dc_ignUN9zG:-LU9zh4D7XF3e5Q_MD5_:bvctbsk) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/b0d4612fd9808a9737b92bc1b4c49b0741b4375a/src/relay/local.rs#L68)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

All the code in the branches could be one level less indented if we do `if plugin.is_empty() {} else {}` instead of a match. This works as well of course, but I think it's more common to just use `if` if the expression is simply a boolean.
</blockquote></details>

Certainly. I'm not sure what made me choose `match`.

**Review by faern — COMMENTED** [review]
> Reviewed 1 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: 12 of 14 files reviewed, 3 unresolved discussions (waiting on @faern and @mvd-ows)

___
*[src/plugin/monitor.rs, line 37 at r4](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU9yxXqETXW-7LFv60Y:-LU9yxXqETXW-7LFv60Z:bl36w8o) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/b0d4612fd9808a9737b92bc1b4c49b0741b4375a/src/plugin/monitor.rs#L37)):*
> ```Rust
>                 }
>             }
>             drop(plugin);
> ```

If you explicitly want this for readability that's fine. Just wanted to highlight that it's probably not technically needed. Since the thread closure is marked as `move ||` and `plugin` is used in the closure, the plugins ownership will be moved to the closure. As such they should be dropped automatically when the closure goes out of scope.
___
*[src/plugin/monitor.rs, line 42 at r4](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LUA-i7S-r_X-dQ4Pv-7:-LUA-i7S-r_X-dQ4Pv-8:b3g26q1) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/b0d4612fd9808a9737b92bc1b4c49b0741b4375a/src/plugin/monitor.rs#L42)):*
> ```Rust
>     }
> 
>     // Wait for any of the plugin processes to exit.
> ```

Minor, maybe write "Wait for the first" to make it clearer that this part is about the first one?
___
*[src/plugin/monitor.rs, line 57 at r4](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LUA0L2X7iZ5c22r19vZ:-LUA0L2X7iZ5c22r19v_:b-rlpiz9) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/b0d4612fd9808a9737b92bc1b4c49b0741b4375a/src/plugin/monitor.rs#L57)):*
> ```Rust
>             .collect()
>             .map(|_| ())
>             .map_err(|_: ()| io::Error::new(io::ErrorKind::Other, "Placeholder"))
> ```

Nit, `_: ()` could be written just `()`. `()` is both a value and a type. So the full version of this would be `(): ()`, but since the compiler can figure out the type of the *value* `()` the `: ()` part can be omitted... I'm pretty sure, have not tested in this very instance.

**Review by mvd-ows — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: 12 of 14 files reviewed, all discussions resolved (waiting on @faern)

___
*[src/plugin/monitor.rs, line 37 at r4](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU9yxXqETXW-7LFv60Y:-LUAEUsGCr0O7B4U6-0W:b-nppg8d) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/b0d4612fd9808a9737b92bc1b4c49b0741b4375a/src/plugin/monitor.rs#L37)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

If you explicitly want this for readability that's fine. Just wanted to highlight that it's probably not technically needed. Since the thread closure is marked as `move ||` and `plugin` is used in the closure, the plugins ownership will be moved to the closure. As such they should be dropped automatically when the closure goes out of scope.
</blockquote></details>

Now that we explicitly terminate the process, there is no reason to keep the `drop` statement. Removed it.
___
*[src/plugin/monitor.rs, line 42 at r4](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LUA-i7S-r_X-dQ4Pv-7:-LUAEnq4-s6VCvek5Ecg:b-896fix) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/b0d4612fd9808a9737b92bc1b4c49b0741b4375a/src/plugin/monitor.rs#L42)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Minor, maybe write "Wait for the first" to make it clearer that this part is about the first one?
</blockquote></details>

Done.
___
*[src/plugin/monitor.rs, line 57 at r4](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LUA0L2X7iZ5c22r19vZ:-LUAEpi78v368tpV8MsG:b-896fix) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/b0d4612fd9808a9737b92bc1b4c49b0741b4375a/src/plugin/monitor.rs#L57)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Nit, `_: ()` could be written just `()`. `()` is both a value and a type. So the full version of this would be `(): ()`, but since the compiler can figure out the type of the *value* `()` the `: ()` part can be omitted... I'm pretty sure, have not tested in this very instance.
</blockquote></details>

Done.
___
*[src/relay/local.rs, line 68 at r4](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LU9qwgD-dc_ignUN9zG:-LUAErRaBHDSJ_EJJ4Sr:b-896fix) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/b0d4612fd9808a9737b92bc1b4c49b0741b4375a/src/relay/local.rs#L68)):*
<details><summary><i>Previously, mvd-ows wrote…</i></summary><blockquote>

Certainly. I'm not sure what made me choose `match`.
</blockquote></details>

Done.

**Review by faern — APPROVED** [review]
> Reviewed 1 of 7 files at r4, 7 of 7 files at r5.
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: all files reviewed, 1 unresolved discussion (waiting on @mvd-ows)

___
*[src/relay/server.rs, line 95 at r5](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LUB4puD0rkOQe4i7HbK:-LUB4puD0rkOQe4i7HbL:bqwgkzw) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/575c8cb90c3ee4e632b876af6675a9807472defa/src/relay/server.rs#L95)):*
> ```Rust
>                 ))),
>             });
>         return boxed_future(f);
> ```

Nit, Both these `return boxed_future(f);`s (here and in the if-branch above) can be simplified to just `boxed_future(f)`. Both of them are the last expressions in the branches and thus they will automatically become the return values of those branches. That would make the entire, large, `if` a statement and as a result making this function more declarative.

**Review by mvd-ows — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: :shipit: complete! all files reviewed, all discussions resolved

___
*[src/relay/server.rs, line 95 at r5](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-LUB4puD0rkOQe4i7HbK:-LUCKvF43al07rZiwv8G:b-896fix) ([raw file](https://github.com/mullvad/shadowsocks-rust/blob/575c8cb90c3ee4e632b876af6675a9807472defa/src/relay/server.rs#L95)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Nit, Both these `return boxed_future(f);`s (here and in the if-branch above) can be simplified to just `boxed_future(f)`. Both of them are the last expressions in the branches and thus they will automatically become the return values of those branches. That would make the entire, large, `if` a statement and as a result making this function more declarative.
</blockquote></details>

Done.

**Review by faern — APPROVED** [review]
> Reviewed 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by faern — APPROVED** [review]
> Reviewed 4 of 7 files at r7.
> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by faern — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/mullvad/shadowsocks-rust/1#-:-LUG6UnLCDDKRGKwi6nx:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request closed without merging**